### PR TITLE
RAVCOM-297

### PR DESCRIPTION
**Pull request by Abhi-210793**

### CODE REVIEW

File: core/src/main/java/org/apache/calcite/rel/rel2sql/RelToSqlConverter.java [hunk 1]
```diff
@@ -593,9 +722,147 @@ private Builder visitAggregate(Aggregate e, List<Integer> groupKeyList,
...
+
+  // Function to update the SELECT list in the subQuery
+  private void updateSelectListInSubQuery(
+      SqlBasicCall sqlBasicCall, List<SqlIdentifier> sqlIdentifiersNew) {
+    List<String> columnNamesInSelect = new ArrayList<>();
+    for (SqlNode sqlNode : sqlBasicCall.getOperands()) {
+      if (sqlNode instanceof SqlSelect) {
+        SqlNodeList sqlNodeList = ((SqlSelect) sqlNode).getSelectList();
+        if (sqlNodeList != null) {
+          for (SqlNode selectNode : sqlNodeList) {
```
Comment (sauajmera) [inline core/src/main/java/org/apache/calcite/rel/rel2sql/RelToSqlConverter.java:818]: this code looks complex can we do refactoring by the way...

private void updateSelectListInSubQuery(SqlBasicCall sqlBasicCall, List<SqlIdentifier> sqlIdentifiersNew) {
    for (SqlNode sqlNode : sqlBasicCall.getOperands()) {
        if (sqlNode instanceof SqlSelect) {
            updateSelectListInSelectNode((SqlSelect) sqlNode, sqlIdentifiersNew);
        }
    }
}

private void updateSelectListInSelectNode(SqlSelect sqlSelect, List<SqlIdentifier> sqlIdentifiersNew) {
    SqlNodeList sqlNodeList = sqlSelect.getSelectList();

    if (sqlNodeList != null) {
        List<String> columnNamesInSelect = getColumnNamesInSelect(sqlNodeList);

        for (SqlIdentifier sqlIdentifier : sqlIdentifiersNew) {
            if (!columnNamesInSelect.contains(sqlIdentifier.names.get(0))) {
                sqlNodeList.add(createNewSelectItem(sqlIdentifier, sqlSelect));
            }
        }

        sqlSelect.setSelectList(new SqlNodeList(sqlNodeList, POS));
    } else if (sqlIdentifiersNew.size() > 0) {
        updateSelectListForJoin(sqlSelect, sqlIdentifiersNew);
    }
}

private List<String> getColumnNamesInSelect(SqlNodeList sqlNodeList) {
    List<String> columnNamesInSelect = new ArrayList<>();

    for (SqlNode selectNode : sqlNodeList) {
        if (selectNode instanceof SqlIdentifier) {
            SqlIdentifier identifier = (SqlIdentifier) selectNode;
            columnNamesInSelect.add(identifier.names.size() > 1 ? identifier.names.get(1) : identifier.names.get(0));
        }
    }

    return columnNamesInSelect;
}

private void updateSelectListForJoin(SqlSelect sqlSelect, List<SqlIdentifier> sqlIdentifiersNew) {
    if (sqlSelect.getFrom() instanceof SqlJoin) {
        SqlNodeList sqlNodes = new SqlNodeList(POS);

        for (SqlIdentifier sqlIdentifier : sqlIdentifiersNew) {
            sqlNodes.add(createNewSelectItem(sqlIdentifier, sqlSelect));
        }

        sqlSelect.setSelectList(new SqlNodeList(sqlNodes, POS));
    }
}

File: core/src/main/java/org/apache/calcite/rel/rel2sql/RelToSqlConverter.java [hunk 2]
```diff
@@ -593,9 +722,147 @@ private Builder visitAggregate(Aggregate e, List<Integer> groupKeyList,
...
+            }
+          }
+        }
+      }
+    }
+  }
+
+  // Function to create a new SqlSelectItem with an alias if it ends with a digit
+  private SqlNode createNewSelectItem(SqlIdentifier sqlIdentifier, SqlNode sqlNode) {
+    if (endsWithDigit(sqlIdentifier.names.get(0))) {
```
Comment (sauajmera) [inline core/src/main/java/org/apache/calcite/rel/rel2sql/RelToSqlConverter.java:850]: Please refactor this code 
'private SqlNode createNewSelectItem(SqlIdentifier sqlIdentifier, SqlNode sqlNode) {
    ImmutableList<String> si = getOperandNames(sqlNode);

    if (endsWithDigit(sqlIdentifier.names.get(0))) {
        return createAsSqlNode(si.get(0), sqlIdentifier.names.get(0));
    } else {
        return new SqlIdentifier(ImmutableList.of(si.get(0), sqlIdentifier.names.get(0)), SqlParserPos.ZERO);
    }
}

private ImmutableList<String> getOperandNames(SqlNode sqlNode) {
    SqlNode[] operands = getOperandsFromJoin(sqlNode);
    return ((SqlIdentifier) operands[1]).names;
}

private SqlNode[] getOperandsFromJoin(SqlNode sqlNode) {
    if (sqlNode instanceof SqlSelect) {
        SqlJoin joinNode = (SqlJoin) ((SqlSelect) sqlNode).getFrom();
        SqlBasicCall basicCall = (SqlBasicCall) (joinNode.getRight() != null ? joinNode.getRight() : joinNode.getLeft());
        return basicCall.getOperands();
    }
    return null;
}

private SqlNode createAsSqlNode(String alias, String columnName) {
    return SqlStdOperatorTable.AS.createCall(POS,
            ImmutableList.of(new SqlIdentifier(ImmutableList.of(alias, columnName.replaceAll("\\d+$", "")), SqlParserPos.ZERO),
                    new SqlIdentifier(columnName, SqlParserPos.ZERO)));
}
'

File: core/src/main/java/org/apache/calcite/sql2rel/RelDecorrelator.java [hunk 3]
```diff
@@ -996,7 +999,11 @@ private Frame decorrelateInputWithValueGenerator(RelNode rel, Frame frame) {
       }
       // If all correlation variables are now satisfied, skip creating a value
       // generator.
-      if (map.size() == corVarList.size()) {
+      for (CorRef corVar : corVarList) {
+        temp = temp.concat(".").concat(String.valueOf(corVar.field));
```
Comment (sauajmera) [inline core/src/main/java/org/apache/calcite/sql2rel/RelDecorrelator.java:1003]: if we can replace this with stream 
temp = corVarList.stream()
          .map(corVar -> "." + corVar.field)
          .collect(Collectors.joining());

Comment (Abhi-210793) [reply core/src/main/java/org/apache/calcite/sql2rel/RelDecorrelator.java]: Done

File: core/src/main/java/org/apache/calcite/rel/rel2sql/RelToSqlConverter.java (hunk 2 above)
Comment (Abhi-210793) [reply core/src/main/java/org/apache/calcite/rel/rel2sql/RelToSqlConverter.java]: Done

File: core/src/main/java/org/apache/calcite/rel/rel2sql/RelToSqlConverter.java (hunk 1 above)
Comment (Abhi-210793) [reply core/src/main/java/org/apache/calcite/rel/rel2sql/RelToSqlConverter.java]: Done

[the remark this answers is not in the document]
Comment (Abhi-210793) [reply core/src/main/java/org/apache/calcite/rel/rel2sql/RelToSqlConverter.java]: Done

File: core/src/main/java/org/apache/calcite/rel/rel2sql/RelToSqlConverter.java [hunk 4]
```diff
@@ -261,7 +264,133 @@ public Result visit(Join e) {
...
+    return false;
+  }
+
+  private List<String> getColumnsUsedInOnConditionWithSubQueryAlias(
+      SqlNode sqlCondition, String neededAlias) {
+    List<String> columnsUsed = new ArrayList<>();
+    List<SqlIdentifier> sqlIdentifierList =
+        collectSqlIdentifiers(((SqlBasicCall) sqlCondition).getOperands());
+    for (SqlIdentifier sqlIdentifier : sqlIdentifierList) {
+      if (sqlIdentifier.names.get(0).equals(neededAlias)) {
```
[the remark this answers is not in the document]
Comment (Abhi-210793) [reply core/src/main/java/org/apache/calcite/rel/rel2sql/RelToSqlConverter.java]: Done

File: core/src/main/java/org/apache/calcite/rel/rel2sql/RelToSqlConverter.java [hunk 5]
```diff
@@ -261,7 +264,133 @@ public Result visit(Join e) {
...
+
+  private void updateResultSelectList(Result leftResult, List<SqlNode> sqlIdentifierList) {
+    if (sqlIdentifierList.size() >= 1 && leftResult.node instanceof SqlSelect) {
+      ((SqlSelect) leftResult.node).setSelectList(new SqlNodeList(sqlIdentifierList, POS));
+    }
+  }
+
+
+  private static boolean endsWithDigit(String str) {
+    if (str.length() > 0) {
```
[the remark this answers is not in the document]
Comment (Abhi-210793) [reply core/src/main/java/org/apache/calcite/rel/rel2sql/RelToSqlConverter.java:364]: Already used

File: core/src/main/java/org/apache/calcite/rel/rel2sql/RelToSqlConverter.java [hunk 6]
```diff
@@ -261,7 +264,133 @@ public Result visit(Join e) {
...
+    return new SqlIdentifier(ImmutableList.of(columnName), POS);
+  }
+
+  private boolean isJoinNodeBasicCall(Result result) {
+    return result.node instanceof SqlSelect && ((SqlSelect) result.node).getFrom() instanceof SqlJoin &&
+        (((SqlJoin) ((SqlSelect) result.node).getFrom()).getLeft()) instanceof SqlBasicCall;
+  }
+
+  private SqlNode createAsSqlIdentifierForColumn(Result leftResult, String columnName) {
+    SqlJoin sqlJoin = ((SqlJoin) ((SqlSelect) leftResult.node).getFrom());
```
[the remark this answers is not in the document]
Comment (Abhi-210793) [reply core/src/main/java/org/apache/calcite/rel/rel2sql/RelToSqlConverter.java]: Done

File: core/src/main/java/org/apache/calcite/rel/rel2sql/RelToSqlConverter.java [hunk 7]
```diff
@@ -261,7 +264,133 @@ public Result visit(Join e) {
...
+        && sqlCondition.toString().contains("`" + result.neededAlias + "`")
+        && result.asSelect().getSelectList() == null;
+  }
+
+  private SqlNode createSqlIdentifierForColumn(Result result, String columnName) {
+    if (endsWithDigit(columnName) && result.node instanceof SqlSelect) {
+      return createAsSqlIdentifierForColumn(result, columnName);
+    } else {
+      if (isJoinNodeBasicCall(result)) {
+        SqlBasicCall sqlBasicCall = ((SqlBasicCall) ((SqlJoin) ((SqlSelect) result.node).getFrom()).getLeft());
```
[the remark this answers is not in the document]
Comment (Abhi-210793) [reply core/src/main/java/org/apache/calcite/rel/rel2sql/RelToSqlConverter.java]: Done

File: core/src/main/java/org/apache/calcite/rel/rel2sql/RelToSqlConverter.java [hunk 8]
```diff
@@ -261,7 +264,133 @@ public Result visit(Join e) {
...
+      }
+        if(result.node instanceof SqlSelect && ((SqlSelect) result.node).getFrom() instanceof SqlJoin) {
+          updateResultSelectList(result, sqlIdentifierList);
+        }
+    }
+  }
+
+  private boolean shouldHandleResultAlias(Result result, SqlNode sqlCondition) {
+    return result.neededAlias != null
+        && sqlCondition.toString().contains("`" + result.neededAlias + "`")
```
[the remark this answers is not in the document]
Comment (Abhi-210793) [reply core/src/main/java/org/apache/calcite/rel/rel2sql/RelToSqlConverter.java]: Done

File: core/src/main/java/org/apache/calcite/rel/rel2sql/RelToSqlConverter.java (hunk 5 above)
[the remark this answers is not in the document]
Comment (sauajmera) [reply core/src/main/java/org/apache/calcite/rel/rel2sql/RelToSqlConverter.java:364]: check if we can make it single liner as below 

return !str.isEmpty() && Character.isDigit(str.charAt(str.length() - 1));

**Comment by sauajmera** [comment]
Please check these two methods both looks so similar , if we can reuse it and make it one.
**findSingleIdentifierInOperands** & **findSingleIdentifier**

File: core/src/main/java/org/apache/calcite/rel/rel2sql/RelToSqlConverter.java (hunk 5 above)
[the remark this answers is not in the document]
Comment (Abhi-210793) [reply core/src/main/java/org/apache/calcite/rel/rel2sql/RelToSqlConverter.java]: Done

**Review by sauajmera — APPROVED** [review]
Changes looks good.

**Comment by sauajmera** [comment]
Changes looks good to me.

**Review by vijayjogi-dm — APPROVED** [review]
Ok with changes

**Pull request merged**